Add jenkins->jira integration script

jmckenzie-dev opened the pull request:
There's a couple of TODO in there but they shouldn't block our initial roll out. What I envision happening here:

1) For a couple weeks, I'll run this manually daily against builds on Jenkins and update the respective CASSANDRA-NNNNN tickets with the results of this script after ironing out kinks
2) Once we feel confident that the script is working consistently and can go unattended, we cross the bridge of how to get there (TODO for account, TODO for depth requirement, etc)

I've run mypy against this; 1 error comes up about imports lacking typing stubs - [not too worried there](https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports).

pylint has more to say about things but I've fixed all the things I think are worth fixing at the moment. Reviewers may disagree.

jmckenzie-dev committed:
Add jenkins->jira integration script

This script will parse a jenkins job of test failures and format a comment to
be added to JIRA about the build run, test failures, and attempt to link to
existing JIRA tickets around those failures. Functionality exists in the script
(as yet untested) to automate updating tickets with the results of this
parsing; we'll need to determine where this is going to run in what contexts as
well as set up an account for it to post from (JenkinsBot?) and then test and
commit that change in tandem.

michaelsembwever commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+jira
+jenkins
```
what's this file doing at the top-level? 
if required by `jenkins_jira_integration.py` maybe both files can go into their own folder, e.g. `cassandra-builds/jenkins-jira-integration/`

michaelsembwever commented on jenkins-dsl/jenkins_jira_integration.py:
```diff
@@ -0,0 +1,352 @@
+    log('   Hit limit of builds. Saving JSON.')
+    with open(json_file, 'w', encoding='utf-8') as outfile:
+        log('Updating the cache with recently queried results')
+        json.dump(cached_data, outfile)
```
Great that the json is a small subset of the original data. Do we have an idea of its actual size?

We have consistent problems with diskspace on the jenkins agents, if this script is intended later to be run as part of jenkin's builds. The cache would be created on every jenkins agent, and the script can be called concurrently (as there is configured two executors on each agent).

Is there any value in centralising the cache on nightlies.a.o (to avoid 40 agents all pulling the original json api files)?

jmckenzie-dev replied:
All the trunk builds we have in history are roughly 50k I believe. So I think we should be fine on space.

dcapwell replied:
gzip compresses well, we can always gzip the file to shrink this further

dcapwell commented on jenkins-dsl/jenkins_jira_integration.py:
```diff
@@ -0,0 +1,352 @@
+
+    required.add_argument('--jenurl', type=str, help='Jenkins server url')
+    required.add_argument('--jenuser', type=str, required=True, help='JIRA server url')
+    required.add_argument('--jenpass', type=str, required=True, help='JIRA username')
```
rather than pass in, can we read from a file?  Better security as command lines can be logged in plain text on the box running this command

jmckenzie-dev replied:
I've been running a shell script with credentials in it scoped to just my user account so things don't hit the history file at least. Is that materially different than parsing credentials from a file?

i.e. having something like `linkJenkinsToJira.sh <build_num>` that has the credentials stored in it scoped to just the owner of the file re: system permissions.

I thought about this a bit and, outside linux permissions on files, couldn't think of a way to cron something that required credentials without having some way to unwind them (key in .py script, plain text in file, etc). So it seemed like this was just as secure a solution as the alternatives, but that's depending on what you can pull from history with a shell script I suppose?

dcapwell commented on jenkins-dsl/jenkins_jira_integration.py:
```diff
@@ -0,0 +1,352 @@
+    required.add_argument('--jenpass', type=str, required=True, help='JIRA username')
+    required.add_argument('--jiraurl', type=str, required=True, help='JIRA server url')
+    required.add_argument('--jirauser', type=str, required=True, help='JIRA username')
+    required.add_argument('--jirapass', type=str, required=True, help='JIRA password')
```
same, can we load from a file rather than from CLI?

dcapwell commented on jenkins-dsl/jenkins_jira_integration.py:
```diff
@@ -0,0 +1,352 @@
+        ser/deser anyway, so best not to mix.
+    :return: testnu->
+    """
+    if not os.path.isdir('cache'):
+        os.mkdir('cache')
```
should this not be a config or some non-relative place?  If run in a job then the cache will be in the workspace, which gets wiped each run.

jmckenzie-dev replied:
Maybe. I originally wrote this with the model of it running from a single place repeatedly rather than on jenkins agents that get wiped. If we go the agent route, we also have to wrestle with the fact that each would have their own local history to execute from, meaning it would be quite difficult to build a long-running history for each branch.

dcapwell replied:
in oss do we have fixed boxes to run this as a cronjob?  the "cache" aspect kinda depends on that

dcapwell commented on jenkins-dsl/jenkins_jira_integration.py:
```diff
@@ -0,0 +1,352 @@
+        else:
+            print('   Did not find build_number: ' + buildnum + ' in the json cache. Populating...')
+            # Pull the data from the Jenkins server as we don't have it in our cache
+            build_data = BuildData(server, branch, buildnum)
+
+            # The data in our JSON cache is a pretty simple subset of what we have in BuildData for a full CI report
+            newcache_data = {}
+            newcache_data[NUMBER] = build_data.number
+            newcache_data[PREVIOUS_NUMBER] = build_data.previous_number
+            # Can't serialize sets to JSON
+            newcache_data['failures'] = list(build_data.test_failures)
```
does this not include uncompleted builds?  Do we wish to cache non-completed builds?

jmckenzie-dev replied:
This includes both failed builds and incomplete builds. Since we're not doing a N vs. N-1 comparison and instead comparing against the entire history, the downside to this is our N of M test failure calculation will also include failed builds in it.

Are they frequent enough we expect this to heavily throw off the visual ability to, at a glance, get an intuition of whether a test is a recent failure (1 of M) or intermittent (N of M)?

dcapwell commented on jenkins-dsl/jenkins_jira_integration.py:
```diff
@@ -0,0 +1,352 @@
+def retrieve_build_details(server: jenkins, branch: str, build_num: int) -> BuildData:
+    result = None
+    found_build = False
+    while not found_build:
+        log('Retrieving requested build: ' + str(build_num) + '. Branch: ' + branch + '. build_num: ' + str(build_num))
+        result = BuildData(server, branch, str(build_num))
+
+        # Some commit messages aren't JIRA related (ninjas, deb release changes, etc)
+        if result.JIRA == UNKNOWN:
+            print('No related CASSANDRA-NNNNN Jira found for build: ' + str(result.number) + '. Commit Message: ' + result.commit_msg)
+            print('Exiting processing; nothing to be done for this build if we can\'t determine the JIRA ticket it\'s associated with.')
+            sys.exit(-1)
+
+        log('Parsed JIRA number: ' + result.JIRA + ' for build: ' + str(result.number))
+        break
```
is this not `while True`?  I don't see you updating `found_build`

jmckenzie-dev replied:
Hah. Vestige of prior impl. Fixing. :+1:

jmckenzie-dev replied:
Just removed loop entirely; used to walk until I found next valid build to compare against as initial model was 1:1 comparison of build vs. last. Now do history, so this method is really about just parsing this one build and seeing if we can infer Jira from it.

dcapwell commented on jenkins-dsl/jenkins_jira_integration.py:
```diff
@@ -0,0 +1,352 @@
+    fail_count = 0
+    total = 0
+
+    for build in data:
```
you can also do `for build, build_data in data.items()`

dcapwell commented on jenkins-dsl/jenkins_jira_integration.py:
```diff
@@ -0,0 +1,352 @@
+    query = 'project = CASSANDRA and resolution = unresolved and summary ~ "' + short_name + '"'
+    log('About to query via the following JQL: ' + query)
+    try:
+        has_jira = jira.search_issues(query)
```
does this do retries?

jmckenzie-dev replied:
As in, mutate the JQL to try with a different form, or gracefully deal with the fact that sometimes Jira won't reply?

In the event Jira doesn't reply, the link it generates will be to the parent kanban board so theoretically a dev would click on that, see the ticket, and not create a new one. This is assuming folks even click these links... 😢

michaelsembwever commented on jenkins-dsl/jenkins_jira_integration.py:
```diff
@@ -0,0 +1,355 @@
+
+        self.paths = []
+        # TODO: Determine if depth=5 is necessary or in any way meaningful in this API Query. If so, document why.
+        build = server.get_build_info('Cassandra-' + branch, int(build_num), depth=5)
```
Does the python api not have selectors like xml xpath or json tree? (you don't need to be pulling out 50MB+ each time…)

jmckenzie-dev replied:
I didn't see it: https://python-jenkins.readthedocs.io/en/latest/api.html#jenkins.Jenkins.get_build_info

There's the ability to pass a tree property w/specific things you want from Jenkins but I didn't see that support in the API reference for this lib. ([reference](https://ci-cassandra.apache.org/job/Cassandra-trunk/970/testReport/api/))

The downside is a big fetch on initial history build + 50MB fetch every N hours as Jenkins jobs complete. Might be worth optimizing to query the API directly instead of leaning on this python wrapper for this query here, but I didn't want to dig into that in V1.

Probably worth a follow up ticket.

jmckenzie-dev commented:
Ok, revisiting this now that we're in the freeze. I've rebased, squashed, and also hard over-ridden the "update jira automatically" part of the script so it'll be manual until we determine where it'll live and what user account it'll run from.

@michaelsembwever and @dcapwell - we good to merge this? It's well isolated and I've been running it by hand on my laptop for a few months now.

michaelsembwever commented:
+1
